File: src/main/java/org/cascadebot/cascadebot/data/objects/GuildSettingsCore.java
/*
 * Copyright (c) 2019 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.data.objects;

import com.google.common.collect.Sets;
import jdk.jfr.Name;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.cascadebot.cascadebot.commandmeta.Module;
import org.cascadebot.cascadebot.commandmeta.ModuleFlag;
import org.cascadebot.cascadebot.data.Config;

import java.lang.reflect.Field;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

@SettingsContainer(module = Module.CORE)
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class GuildSettingsCore {

    public static Map<String, Field> VALUES = new HashMap<>();

    private long guildId;

    static {
        for (Field field : GuildSettingsCore.class.getDeclaredFields()) {
            if (field.getName().equals("VALUES")) continue;
            if (field.getAnnotation(Setting.class) == null || !field.getAnnotation(Setting.class).directlyEditable()) continue;
            field.setAccessible(true);
            VALUES.put(field.getName().toLowerCase(), field);
        }
    }

    //region Boolean flags
    @Setting
    private boolean mentionPrefix = false; // Whether the bot will respond to a mention as a prefix

    @Setting
    private boolean deleteCommand = true;

    @Setting
    private boolean useEmbedForMessages = true;

    @Setting
    private boolean showPermErrors = true; // Whether commands will silently fail on no permissions

    @Setting
    private boolean showModuleErrors = true;

    @Setting
    private boolean adminsHaveAllPerms = true;

    @Setting
    private boolean allowTagCommands = true; // Whether tag commands will be executed by ;<tagname>

    @Setting
    private boolean helpHideCommandsNoPermission = true;

    @Setting
    private boolean helpShowAllModules = false;

    //endregion

    @Setting(directlyEditable = false)
    private Set<Module> enabledModules = Sets.newConcurrentHashSet(Module.getModules(ModuleFlag.DEFAULT));

    @Setting(directlyEditable = false)
    private Set<CommandFilter> commandFilters = Sets.newConcurrentHashSet();

    @Setting(directlyEditable = false)
    private String prefix = Config.INS.getDefaultPrefix();

    @Setting(directlyEditable = false)
    private ConcurrentHashMap<String, Tag> tags = new ConcurrentHashMap<>();

    public GuildSettingsCore(long guildId) {
        this.guildId = guildId;
    }

    //region Modules
    public boolean enableModule(Module module) {
        if (module.isPrivate()) {
            throw new IllegalArgumentException("This module is not available to be enabled!");
        }
        return this.enabledModules.add(module);
    }

    public boolean disableModule(Module module) {
        if (module.isPrivate()) {
            throw new IllegalArgumentException("This module is not available to be disabled!");
        } else if (module.isRequired()) {
            throw new IllegalArgumentException(String.format("Cannot disable the %s module!", module.toString().toLowerCase()));
        }
        return this.enabledModules.remove(module);
    }

    public boolean isModuleEnabled(Module module) {
        boolean isEnabled = this.enabledModules.contains(module);
        if (!isEnabled && module.isRequired()) {
            this.enabledModules.add(module);
            return true;
        }
        return isEnabled;
    }

    //endregion

    public Map<String, Tag> getTags() {
        return Collections.unmodifiableMap(tags);
    }

    public Tag getTag(String key) {
        return tags.get(key);
    }

    public boolean hasTag(String key) {
        return tags.containsKey(key);
    }

    public void addTag(String key, Tag tag) {
        tags.put(key, tag);
    }

    public boolean removeTag(String key) {
        return tags.remove(key) != null;
    }

    public Set<CommandFilter> getCommandFilters() {
        return Collections.unmodifiableSet(commandFilters);
    }

    public boolean hasCommandFilter(String name) {
        return getCommandFilter(name) != null;
    }

    public CommandFilter getCommandFilter(String name) {
        return commandFilters.stream()
                .filter(filter -> filter.getName().equals(name))
                .findFirst()
                .orElse(null);
    }

    public boolean addCommandFilter(CommandFilter filter) {
        return commandFilters.add(filter);
    }

    public boolean removeCommandFilter(String name) {
        return commandFilters.removeIf(filter -> filter.getName().equals(name));
    }

}

File: src/main/kotlin/org/cascadebot/cascadebot/data/objects/GuildSettingsCore.kt
package org.cascadebot.cascadebot.data.objects

import com.google.common.collect.Sets
import org.bson.BsonDocument
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.commandmeta.MainCommand
import org.cascadebot.cascadebot.commandmeta.Module
import org.cascadebot.cascadebot.data.Config
import org.cascadebot.cascadebot.data.database.BsonObject
import org.cascadebot.cascadebot.data.database.DataHandler
import org.cascadebot.cascadebot.data.language.Locale
import org.cascadebot.cascadebot.utils.GuildDataUtils.assertWriteMode
import org.cascadebot.cascadebot.utils.ifContains
import org.cascadebot.cascadebot.utils.ifContainsArray
import org.cascadebot.cascadebot.utils.ifContainsBoolean
import org.cascadebot.cascadebot.utils.ifContainsString
import java.lang.UnsupportedOperationException
import java.util.concurrent.ConcurrentHashMap


@SettingsContainer(module = Module.CORE)
class GuildSettingsCore : BsonObject {

    @Setting
    var mentionPrefix = false

    @Setting
    var deleteCommand = true

    @Setting
    var useEmbedForMessages = true

    @Setting
    var showPermErrors = true

    @Setting
    var showModuleErrors = true

    @Setting
    var adminsHaveAllPerms = true

    @Setting
    var helpHideCommandsNoPermission = true

    @Setting
    var helpShowAllModules = false

    var locale: Locale = Locale.getDefaultLocale()
    var prefix: String = Config.INS.defaultPrefix

    private val commandInfo = ConcurrentHashMap<Class<MainCommand>, MutableSet<GuildCommandInfo>>()
    private val enabledModules: MutableSet<Module> = Sets.newConcurrentHashSet(Module.getModules(ModuleFlag.DEFAULT))

    

    //region Modules
    fun enableModule(module: Module): Boolean {
        assertWriteMode()
        require(!module.isPrivate) { "This module is not available to be enabled!" }
        return enabledModules.add(module)
    }

    fun disableModule(module: Module): Boolean {
        assertWriteMode()
        require(!module.isPrivate) { "This module is not available to be disabled!" }
        require(!module.isRequired) { "Cannot disable the ${module.toString().toLowerCase()} module!" }
        return enabledModules.remove(module)
    }

    fun isModuleEnabled(module: Module): Boolean {
        val isEnabled = enabledModules.contains(module)
        if (!isEnabled && module.isRequired) {
            enabledModules.add(module)
            return true
        }
        return isEnabled
    }

    //endregion

    //region Commands
    fun enableCommand(command: MainCommand) {
        assertWriteMode()
        if (command.module().isPrivate) return
        if (commandInfo.contains(command.javaClass)) {
            getGuildCommandInfo(command).enabled = true
        }
    }

    fun enableCommandByModule(module: Module) {
        assertWriteMode()
        if (module.isPrivate) return
        for (command in CascadeBot.INS.commandManager.getCommandsByModule(module)) {
            enableCommand(command)
        }
    }

    fun disableCommand(command: MainCommand) {
        assertWriteMode()
        if (command.module().isPrivate) return
        getGuildCommandInfo(command).enabled = false
    }

    fun disableCommandByModule(module: Module) {
        assertWriteMode()
        if (module.isPrivate) return
        for (command in CascadeBot.INS.commandManager.getCommandsByModule(module)) {
            disableCommand(command)
        }
    }

    fun isCommandEnabled(command: MainCommand): Boolean {
        return if (commandInfo.contains(command.javaClass)) {
            getGuildCommandInfo(command).enabled
        } else isModuleEnabled(command.module())
    }

    fun getCommandAliases(command: MainCommand): Set<String> {
        return if (commandInfo.contains(command.javaClass)) {
            getGuildCommandInfo(command).aliases.applyChanges(command.globalAliases(locale))
        } else command.globalAliases(locale)
    }

    fun addAlias(command: MainCommand, alias: String): Boolean {
        assertWriteMode()
        return getGuildCommandInfo(command).aliases.add(alias)
    }

    fun removeAlias(command: MainCommand, alias: String): Boolean {
        assertWriteMode()
        return getGuildCommandInfo(command).aliases.remove(alias)
    }

    private fun getGuildCommandInfoSet(command: MainCommand): MutableSet<GuildCommandInfo> {
        return commandInfo.computeIfAbsent(command.javaClass) {
            mutableSetOf(GuildCommandInfo(command, locale))
        }
    }

    private fun getGuildCommandInfo(command: MainCommand, locale: Locale = this.locale): GuildCommandInfo {
        return getGuildCommandInfoSet(command).find { it.locale == locale } ?: run {
            val commandInfo = GuildCommandInfo(command, locale)
            getGuildCommandInfoSet(command).add(commandInfo)
            commandInfo
        }
    }

    //endregion

    override fun fromBson(bsonDocument: BsonDocument) {
        bsonDocument.ifContainsBoolean("mentionPrefix") { mentionPrefix = it }
        bsonDocument.ifContainsBoolean("deleteCommand") { deleteCommand = it }
        bsonDocument.ifContainsBoolean("useEmbedForMessages") { useEmbedForMessages = it }
        bsonDocument.ifContainsBoolean("showPermErrors") { showPermErrors = it }
        bsonDocument.ifContainsBoolean("showModuleErrors") { showModuleErrors = it }
        bsonDocument.ifContainsBoolean("adminsHaveAllPerms") { adminsHaveAllPerms = it }
        bsonDocument.ifContainsBoolean("helpHideCommandsNoPermission") { helpHideCommandsNoPermission = it }
        bsonDocument.ifContainsBoolean("helpShowAllModules") { helpShowAllModules = it }
        bsonDocument.ifContainsString("locale") { locale = Locale.valueOf(it) }
        bsonDocument.ifContainsString("prefix") { prefix = it }
        bsonDocument.ifContainsArray("enabledModules") { array ->
            enabledModules.clear();
            array.map{ it.asString().value } // Convert each array value to a string
                .map { string -> Module.valueOf(string) } // Map each value to a Module enum
                .forEach { enabledModules.add(it) }
        }
    }

    override fun handleRemove(tree: DataHandler.RemovedTree) {

    }

}